frontend_tests/puppeteer_tests/07-navigation.js

```diff
@@ -0,0 +1,106 @@
+    await page.click(menu_selector);
+}
+
+async function navigate_to_settings(page) {
+    console.log("Navigating to settings");
+
+    await open_menu(page);
+
+    const settings_selector = "a[href^='#settings']";
+    await page.waitForSelector(settings_selector, {visible: true});
+    await page.click(settings_selector);
+
+    await page.waitForSelector("#settings_page", {visible: true});
+
+    await page.click("#settings_page .content-wrapper .exit");
+}
+
+async function navigate_to_subscriptions(page) {
+    console.log("Navigate to subscriptions");
+
+    await open_menu(page);
+
+    const manage_streams_selector = 'a[href^="#streams"]';
+    await page.waitForSelector(manage_streams_selector, {visible: true});
+    await page.click(manage_streams_selector);
+
+    await page.waitForSelector("#subscription_overlay", {visible: true});
+    await page.waitForSelector("#subscriptions_table", {visible: true});
+
+    await page.click("#subscription_overlay .exit");
+}
+
+async function test_reload_hash(page) {
+    const initial_page_load_time = await page.evaluate(() => page_params.page_load_time);
+    console.log("initial load time: " + initial_page_load_time);
+
+    const initial_hash = await page.evaluate(() => window.location.hash);
+
+    await Promise.all([
+        page.evaluate(() => reload.initiate({immediate: true})),
+
+        // navigations during reload were causing failures,
+        // this fixes it.
+        page.waitForNavigation({waitUntil: "networkidle0"}),
+    ]);
+
+    const page_load_time = await page.evaluate(() => page_params.page_load_time);
+    assert(page_load_time > initial_page_load_time, "Page not reloaded.");
+
+    const hash = await page.evaluate(() => window.location.hash);
+    assert.strictEqual(hash, initial_hash, "Hash not preserved.");
+}
+
+async function navigation_tests(page) {
+    await common.log_in(page);
+
+    await navigate_to_settings(page);
+
+    const verona_id = await page.evaluate(() => stream_data.get_stream_id("Verona"));
+    const verona_narrow = `narrow/stream/${verona_id}-Verona`;
+
+    await navigate_to(page, verona_narrow, "message_feed_container");
+
+    // Hardcoded this instead of using `navigate_to`
+    // as puppeteer cannot click hidden elements.
+    await page.evaluate(() => $("a[href='#message_feed_container]'").click());
+    await wait_for_tab(page, "message_feed_container");
+
+    await navigate_to_subscriptions(page);
+    await navigate_to(page, "", "message_feed_container");
+    await navigate_to_settings(page);
+    await navigate_to(page, "narrow/is/private", "message_feed_container");
+    await navigate_to_subscriptions(page);
+    await navigate_to(page, verona_narrow, "message_feed_container");
+
+    await test_reload_hash(page);
+
+    // Verify that we're narrowed to the target stream
+    await page.waitForSelector("#zfilt", {visible: true});
+    await page.waitForSelector("#zhome", {visible: false});
```

Comment:
Should this be `visible` or `hidden: true` given 33e19fa7d19ddb201f95c0ee8be34203d9848475?  @sahil839 FYI.

Reply:
In any case, we should probably add a check of the navbar area to confirm that we're seeing the right stream, at least if the comment is claiming that's what we're checking.  (This just checks we're narrowed).

Reply:
This should be `hidden: true`

Reply:
Yes, it should be `hidden: true`. changed it.

Reply:
Removed these `waitFor`'s and added `await common.wait_for_text(page, "#message_view_header .stream", "Verona");` as it is  much better check as you said, Tim.